core/src/main/java/org/apache/iceberg/CatalogUtil.java

```diff
@@ -117,4 +120,39 @@ private static void deleteFiles(FileIO io, Set<ManifestFile> allManifests) {
           }
         });
   }
+
+  /**
+   * Load a custom {@link FileIO} implementation.
+   * The implementation must have a no-arg constructor.
+   * If the class implements {@link org.apache.hadoop.conf.Configurable}, we will also set the Hadoop configuration.
+   * @param impl full class name of a custom FileIO implementation
+   * @param hadoopConf hadoop configuration
+   * @return FileIO class
+   * @throws IllegalArgumentException if class path not found or
+   *  right constructor not found or
+   *  the loaded class cannot be casted to the given interface type
+   */
+  public static FileIO loadCustomFileIO(String impl, Configuration hadoopConf) {
+    LOG.info("Loading custom FileIO implementation: {}", impl);
+    DynConstructors.Ctor<FileIO> ctor;
+    try {
+      ctor = DynConstructors.builder(FileIO.class)
+          .impl(impl)
+          .buildChecked();
+    } catch (NoSuchMethodException e) {
+      throw new IllegalArgumentException(String.format(
+          "Cannot initialize FiloIO, please make sure %s has a no-arg constructor", impl), e);
+    }
+    try {
+      FileIO fileIO = ctor.newInstance();
+      if (fileIO instanceof Configurable) {
+        ((Configurable) fileIO).setConf(hadoopConf);
+      }
+      return fileIO;
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(
+          String.format("Cannot initialize FileIO, fail to cast %s to class %s.",
+              impl, FileIO.class), e);
```

Comment:
I think it would be more clear to say `%s does not implement FileIO`. Passing the class calls `toString`, not `getName`, and that output includes "interface" or "class". This would actually log "fail to cast com.example.X to class interface ...". The problem is that `com.example.X` needs to implement `FileIO`, so saying it directly is a bit better.